src/time_bucket.c

```diff
@@ -16,24 +16,35 @@
 #include <utils/fmgrprotos.h>
 #endif
 
-#define TIME_BUCKET(period, timestamp, min, result)			\
+#define TIME_BUCKET(period, timestamp, offset, min, max, result)			\
 	do															\
 	{															\
 		if (period <= 0) \
 			ereport(ERROR, \
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), \
 				 errmsg("period must be greater then 0"))); \
-		*(result) = (timestamp / period) * period;				\
-		if (timestamp < 0)										\
-			if (timestamp % period)								\
-			{													\
-				if (*(result) < min + period)				\
-					ereport(ERROR, \
-						(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE), \
-						 errmsg("timestamp out of range"))); \
-				else											\
-					*(result) = *(result) - period;				\
-			}													\
+    if (offset)  \
+    { \
+      /* when offset is positive check for underflow, otherwise check for overflow */ \
```

Comment:
I think we need a little more detail on this comment: `We need to ensure that the timestamp is in range _after_ the offset is applied: when the offset is positive we need to make sure it's at least resultant time is at least min, and when negative that it is less than the max`

Reply:
Do we need to check the other direction also? That a positive offset doesn't overflow?